client/src/connection/itc/index.ts

```diff
@@ -0,0 +1,387 @@
+    if (/Setup error/.test(msg)) {
+      this._shellProcess.kill();
+      this._workDirectory = undefined;
+    }
+  };
+
+  /**
+   * Handles stdout output from the powershell child process.
+   * @param data a buffer of stdout output from the child process.
+   */
+  private onShellStdOut = (data: Buffer): void => {
+    const output = data.toString().trimEnd();
+    const outputLines = output.split(/\n|\r\n/);
+
+    outputLines.forEach((line: string) => {
+      if (!line) {
+        return;
+      }
+      if (!this._workDirectory && line.startsWith("WORKDIR=")) {
+        const parts = line.split("WORKDIR=");
+        this._workDirectory = parts[1].trim();
+        this._runResolve();
+        return;
+      }
+      if (line.endsWith(endCode)) {
+        // run completed
+        this.fetchResults();
+      } else {
+        this._html5FileName =
+          line.match(/NOTE: .+ HTML5.* Body.+: (.+)\.htm/)?.[1] ??
+          this._html5FileName;
+        this._onLogFn?.([{ type: "normal", line }]);
+      }
+    });
+  };
+
+  /**
+   * Generic call for use on stdin write completion.
+   * @param err The error encountered on the write attempt. Undefined if no error occurred.
+   */
+  private onWriteComplete = (err: Error): void => {
+    if (err) {
+      this._runReject?.(err);
+    }
+  };
+
+  /**
+   * Not implemented.
+   */
+  public sessionId = (): string => {
+    throw new Error("Not Implemented");
+  };
+
+  /**
+   * Fetches the ODS output results for the latest html results file.
+   */
+  private fetchResults = async () => {
+    const globalStorageUri = getGlobalStorageUri();
+    try {
+      await workspace.fs.readDirectory(globalStorageUri);
+    } catch (e) {
+      await workspace.fs.createDirectory(globalStorageUri);
+    }
+
+    const outputFileUri = Uri.joinPath(
+      globalStorageUri,
+      `${this._html5FileName}.htm`,
+    );
+    this._shellProcess.stdin.write(
+      `$filePath = "${resolve(
+        this._workDirectory,
+        this._html5FileName + ".htm",
+      )}"
+$outputFile = "${outputFileUri.fsPath}"
+$runner.FetchResultsFile($filePath, $outputFile)\n`,
+      this.onWriteComplete,
+    );
+
+    const file = await new Promise((resolve) => {
+      const start = Date.now();
+      const maxTime = 10 * 1000;
+      const interval = setInterval(async () => {
+        try {
+          const file = await workspace.fs.readFile(outputFileUri);
+          clearInterval(interval);
+          resolve(file);
+        } catch (e) {
+          if (Date.now() - maxTime > start) {
+            clearInterval(interval);
+            resolve(null);
+          }
+        }
+      }, 1000);
+    });
+
+    const htmlResults = (file || "").toString();
+    if (file) {
+      workspace.fs.delete(outputFileUri);
+    }
+
+    const runResult: RunResult = {};
+    if (htmlResults.search('<*id="IDX*.+">') !== -1) {
+      runResult.html5 = htmlResults;
+      runResult.title = "Result";
+    }
+    this.storePassword();
```

Comment:
how about store password on session creation success? User may close VS Code before run code for whatever reason.

Reply:
Right now, the two are linked. We don't create the session _until_ the user clicks the button to run the code. And, since run code has the `endCode` to check for, I just relied on that to store the password. I suppose I could do something similar for session creation (since I'm likely going to do the same for fetching results)

Reply:
Resolved in https://github.com/sassoftware/vscode-sas-extension/pull/592/commits/1da05d1a445dde2144bc471b9cc842da3dccebc4